beetsplug/fetchart.py

```diff
@@ -49,66 +58,60 @@
 except ImportError:
     HAS_BEAUTIFUL_SOUP = False
 
-if TYPE_CHECKING:
-    from logging import Logger
-
 
 CONTENT_TYPES = {"image/jpeg": [b"jpg", b"jpeg"], "image/png": [b"png"]}
 IMAGE_EXTENSIONS = [ext for exts in CONTENT_TYPES.values() for ext in exts]
 
 
+class ImageAction(Enum):
+    """Indicates whether an image is useable or requires post-processing."""
+
+    EXACT = 1
+    DOWNSCALE = 2
+    DOWNSIZE = 3
+    DEINTERLACE = 4
+    REFORMAT = 5
+
+
+class MetadataMatch(Enum):
+    """Indicates whether an `Candidate` matches the search criteria exactly."""
+
+    EXACT = 0
+    FALLBACK = 1
+
+
+class SourceLocation(Enum):
+    LOCAL = "local"
+    REMOTE = "remote"
+
+
+@dataclass
 class Candidate:
     """Holds information about a matching artwork, deals with validation of
     dimension restrictions and resizing.
     """
 
-    CANDIDATE_BAD = 0
-    CANDIDATE_EXACT = 1
-    CANDIDATE_DOWNSCALE = 2
-    CANDIDATE_DOWNSIZE = 3
-    CANDIDATE_DEINTERLACE = 4
-    CANDIDATE_REFORMAT = 5
-
-    MATCH_EXACT = 0
-    MATCH_FALLBACK = 1
+    _log: Logger
```

Comment:
Regarding the design: it seems like `Candidate` has no other use except for running validation. The attributes that the object has are only required until the end of the validate method, at which point it becomes `ValidatedCandidate` and becomes either `ExactCandidate` or `InexactCandidate`. A couple of points here

1. It looks to me this should be just a function or a method instead which does not need to hold any state. See the [Middle Man](https://refactoring.guru/smells/middle-man) code smell
2. `Candidate` is the main guy here so I feel like it should be made a bit more relevant by keeping all related logic. 
3. Judging by the class names, the difference between `ExactCandidate` and `InexactCandidate` may not be significant enough to justify the need to create two classes.

All in all, I think the validation looks like a great use case for a [factory method](https://refactoring.guru/design-patterns/factory-method) like
```python
class Candidate:
    @classmethod
    def from_match(cls, _log: Logger, source: ArtSource, ...) -> Self:
        # validate
        ...
        return cls(source=source, path=path)  # return a new Candidate
```

This drops the need for the `ValidatedCandidate` since now `Candidate` must be valid in order to be created.

The `resize` method from `InexactCandidate` can be a method on `Candidate` - which does nothing if the size already meets the expectations.

Reply:
The point of the `ExactCandidate`/`InexactCandidate` split was to avoid assertions in their `.resize()` method (since only for `InexactCandidate` we always know the `size`). I think encoding that using the type system does require different types (or generics).

I'll see what I can achieve by refactoring `Candidate`.

Reply:
> You might be right that this can be simplified: We split the process of finding, downloading, validating and resizing art into separate function, but only ever run them one after each other anyway.

In such case, maybe a couple of `@classmethod`s could be a good _candidate_ (pun intended)?